### DISCUSSION TITLE
Logs show Web Scraping Failed for two sites?

### DISCUSSION THREAD

**Discussion by quella** [FreshRSS/FreshRSS#6004] [Q&A]
Had a question on two different websites which I'm trying to do XPath Web Scraping and in the logs I see a failed with 302, yet I can reach them via a web browser.  I did try the setting to change my user agent, but that does not seem to help.  Any thoughts on why this site below is showing a warning that it failed (302), but it also seems to continue to get the Favicon, etc.

Just want to confirm why some sites may not allow scraping (anti-scrape detection?) and others I can scrape with out any issues.  Am I missing something obvious?  It would also be nice to have a place where people could share their successful scrapes, etc.  Help prevent reinventing the wheel.

**Information Below**

**Freshrss Details:** Latest Development build using Docker running on a RaspberryPI.

**site:** https://www.printables.com/model?o=newest

The other site I was trying to scrape was "Artstation.com"

**Logs:**
FreshRSS[150]: FreshRSS GET html https://www.printables.com/model?o=newest

FreshRSS[150]: [quella] [Sat, 30 Dec 2023 08:25:10 -0500] [warning] --- HTML+XPath Web scraping failed for [https://www.printables.com/model?o=newest]   **<--- WARNING LOG MESSAGE**

192.168.99.178 - - [30/Dec/2023:08:25:10 -0500] "POST /i/?c=feed&a=add HTTP/1.1" 302 - "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36" **<--- 302 Error?**

192.168.99.178 - - [30/Dec/2023:08:25:10 -0500] "GET /i/?c=subscription&a=feed&id=1756&rid=65901a3645c5a HTTP/1.1" 200 7195 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
192.168.99.178 - - [30/Dec/2023:08:25:40 -0500] "POST /i/?c=subscription&a=feed&id=1756 HTTP/1.1" 302 - "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
192.168.99.178 - - [30/Dec/2023:08:25:40 -0500] "GET /i/?c=subscription&id=1756&rid=65901a5497b40 HTTP/1.1" 200 15827 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
FreshRSS[41]: FreshRSS Favicon GET https://www.printables.com/model?o=newest
FreshRSS[41]: FreshRSS Favicon GET https://www.printables.com/
FreshRSS[41]: FreshRSS Favicon GET https://www.printables.com/favicon.ico
192.168.99.178 - - [30/Dec/2023:08:25:40 -0500] "GET /f.php?9a9f71c0 HTTP/1.1" 200 7406 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"

Here is a screenshot of my settings for the above site in the tool:

![Screenshot 2023-12-30 at 8 43 24 AM](https://github.com/FreshRSS/FreshRSS/assets/823352/e02530e4-569e-408a-8943-a4914814ebf4)

**Comment by FebX23** [comment]
Sounds like you're hitting a snag with web scraping, especially dealing with those pesky 302 messages. It's a classic tale of scraper vs. the dynamic web. A 302 status code means you're being redirected, which could happen for a bunch of reasons, including the site trying to send you to a login page, or it might be their way of handling bots (like ours).

Changing the user agent was a smart move, but these sites can be pretty clever with their anti-scraping tech. They might be looking at more than just the user agent—things like your request rate, the absence of typical browser behavior, or even if certain cookies are missing.

For sites like printables.com and Artstation.com, which seem to have more sophisticated detection mechanisms, using a tool like [Crawlbase ](https://crawlbase.com/?utm_source=GitHub_Feb)might give you the edge you need. It's not just about changing your user agent; it helps manage requests in a way that mimics human browsing more closely, which can help avoid detection. Plus, dealing with redirects and cookies becomes a lot simpler.

On sharing successful scrapes, it's a great idea! A community-driven repository or forum where folks share their configs and tips would save tons of time and headaches. Not sure if there's a dedicated spot for this yet, but it would definitely be a valuable resource.

**Comment by Emmeline-Hawthorne** [comment]
It looks like the 302 redirect might be due to anti-scraping measures. Some sites enforce login, block bots, or require JavaScript execution, which can prevent direct scraping.

Here are a few things to try:

Use Session Handling : Maintain cookies and headers to simulate a real user.
Rotate User Agents & Proxies : Some sites flag repeated requests from the same IP.
Check for JavaScript Rendering : If required, use Selenium or Playwright instead.
Inspect the Redirect : Try manually following the 302 URL to see what’s happening.
Let me know if you need more help debugging!
